src/Simulation/QCTraceSimulator/SortedQubitPool.cs

```diff
@@ -0,0 +1,321 @@
+            /// Maximum of all values, that are less than or equal than sample, among values seen after reset.
+            /// </summary>
+            internal ComplexTime MaxLowerBound = ComplexTime.MinValue;
+
+            /// <summary>
+            /// Minimum of all values, that are greater or equal than sample, among values seen after reset.
+            /// </summary>
+            internal ComplexTime MinUpperBound = ComplexTime.MaxValue;
+
+            /// <summary>
+            /// Sample, that will be passed to a series of comparisons.
+            /// Sample is not an object that's stored in the set so it is excluded from the values seen.
+            /// </summary>
+            private QubitTimeNode Sample;
+
+            /// <summary>
+            /// Resets minimum and maximum value of nodes seen from the set.
+            /// Also sets sample object, which should not be counted because it is not an object from the set.
+            /// </summary>
+            /// <param name="sample">Sample object that is not considered for minimum and maximum.</param>
+            internal void ResetForComparison(QubitTimeNode sample) {
+                MaxLowerBound = ComplexTime.MinValue;
+                MinUpperBound = ComplexTime.MaxValue;
+                Sample = sample;
+            }
+
+            /// <summary>
+            /// Compares Time field of two nodes. Also updates MaxLowerBound and MinUpperBound.
+            /// </summary>
+            /// <returns>Result of comparison of Time field of QubitTimeNode</returns>
+            public int Compare(QubitTimeNode a, QubitTimeNode b) {
+                // Note that comparison of a and b nodes to Sample is "by reference" in this function.
+                int result = ComplexTime.Compare(a.Time, b.Time);
+                if (result > 0) {
+                    if (a == Sample && b != Sample) {
+                        MaxLowerBound = ComplexTime.Max(MaxLowerBound, b.Time);
+                    } else if (a != Sample && b == Sample) {
+                        MinUpperBound = ComplexTime.Min(MinUpperBound, a.Time);
+                    }
+                } else if (result < 0) {
+                    if (a == Sample && b != Sample) {
+                        MinUpperBound = ComplexTime.Min(MinUpperBound, b.Time);
+                    } else if (a != Sample && b == Sample) {
+                        MaxLowerBound = ComplexTime.Max(MaxLowerBound, a.Time);
+                    }
+                } else {
+                    // We found the value if one argument is the sample and nother is not.
+                    if ((a == Sample) != (b == Sample)) {
+                        MaxLowerBound = b.Time;
+                        MinUpperBound = b.Time;
+                    }
+                }
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// QubitTimeNodes sorted by Time field. Nodes with the same Time field are stored in a linked list.
```

Comment:
The comment isn't clear. Do you mean, that nodes with same time should _not_ be added to the set but should be tracked using the linked list facility of QubitTimeNode class?

Reply:
It's a standard .net SortedSet, it won't accept nodes with the same time. Updated wording will be in the next push.